api/handlers.go

```diff
@@ -31,3 +32,123 @@ func (api *API) userHandlerGET(w http.ResponseWriter, req *http.Request, _ httpr
+	// TODO Should we create the user in case it's not found?
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	ups, err := api.staticDB.UploadsByUser(req.Context(), *u)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+	}
+	WriteJSON(w, ups)
+}
+
+// userDownloadsHandler returns all downloads made by the current user.
+func (api *API) userDownloadsHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	sub, _, _, err := tokenFromContext(req)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	u, err := api.staticDB.UserBySub(req.Context(), sub)
+	// TODO Should we create the user in case it's not found?
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	ups, err := api.staticDB.DownloadsByUser(req.Context(), *u)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+	}
+	WriteJSON(w, ups)
+}
+
+// trackUploadHandler registers a new upload in the system.
+func (api *API) trackUploadHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	sl := ps.ByName("skylink")
+	sub, _, _, err := tokenFromContext(req)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	skylink, err := api.staticDB.Skylink(req.Context(), sl)
+	if err == database.ErrInvalidSkylink {
+		WriteError(w, err, http.StatusBadRequest)
+		return
+	}
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	u, err := api.staticDB.UserBySub(req.Context(), sub)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	_, err = api.staticDB.UploadCreate(req.Context(), *u, *skylink)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	if skylink.Size == 0 {
+		// Zero size means that we haven't fetched the skyfile's size yet.
+		// Queue the skylink to have its meta data fetched and updated in the
+		// DB, as well as the user's used space to be updated.
+		api.staticMF.Queue <- metafetcher.Message{
+			UserID:    u.ID,
+			SkylinkID: skylink.ID,
+		}
+	} else {
+		err = api.staticDB.UserUpdateUsedStorage(req.Context(), u.ID, skylink.Size)
+		if err != nil {
+			// Log the error but return success - the record will be corrected
+			// later when we rescan the user's used space.
+			api.staticLogger.Debug("Failed to update user's used space:", err)
+		}
+	}
+	WriteSuccess(w)
+}
+
+// trackDownloadHandler registers a new download in the system.
+func (api *API) trackDownloadHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	sl := ps.ByName("skylink")
+	if sl == "" {
+		WriteError(w, errors.New("missing parameter 'skylink'"), http.StatusBadRequest)
+		return
+	}
+	sub, _, _, err := tokenFromContext(req)
+	if err != nil {
+		WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	skylink, err := api.staticDB.Skylink(req.Context(), sl)
```

Comment:
Here we don't seem to care if the Skylink size is 0. I take it that's because on download we don't need to keep track of something like "amount downloaded" like we have to in upload with "used storage"?

Reply:
That's correct.